First draft of FIP-0076: Direct data onboarding (#804)
filecoin-project/FIPs: FIPS/fip-direct-onboarding.md

---
fip: "<to be assigned>"
title: Direct data onboarding
author: "Alex North (@anorth), @zenground0"
discussions-to: https://github.com/filecoin-project/FIPs/discussions/730
status: Draft
type: Technical
category: Core
created: 2023-08-24
---

## Simple Summary

Adds new prove-commit and replica-update methods to the miner actor which support committing to data
and claiming verified allocations without requiring a deal with the built-in market actor.
Adds a sector→deal mapping in the built-in market actor and deprecates the storage of deal IDs in the miner actor.
Removes the legacy Miner::PreCommitSector (method 6) and Miner::PreCommitSectorBatch (method 25),

> in favour of the existing Miner::PreCommitSectorBatch2 (method 28) which requires

Reviewer: Worth mentioning we're replicating the method formerly known as prove replica updates 2 (29) and giving that name to the replica update method being added (34)

  but does not need to specify the structure of that data nor any deals or verified allocations.
- at prove-commit or replica-update, an SP specifies the pieces of data comprising a sector,
  and for each piece may nominate:
    - a verified allocation to claim, and/or
    - an actor to notify of the commitment (e.g. to activate a deal).

The miner actor verifies that the pieces of data claimed by the SP correspond to the data commitment proven.
The verified registry is involved only if verified data is being claimed.
The built-in market actor is involved only if an on-chain-settled deal is required (i.e. if non-zero payment).
Otherwise, the SP simply commits data directly to their sector with no unnecessary overhead.

Existing onboarding methods are retained and remain fully supported, but optional.

### Storage miner actor

No changes to miner state schemas are necessary,
but the deal IDs stored on with each sector's metadata are redundant.
The sector→deal association will be stored in the built-in market actor instead, and
any information that remains in the miner actor state will be incomplete.

The interpretation of some fields changes slightly.

> No migration is necessary.

Reviewer: nit: I would delete this line.  In terms of writing new miner state this is true but all sector infos must be read to make the sector->deal index.  So this FIP requires paying for the most expensive part of any migration.

TODO: remove PreCommitSector and PreCommitSectorBatch methods

#### State

##### SectorPreCommitInfo

The `DealIDs` field remains in use to support existing onboarding methods.
It is required to be empty by the new `ProveCommitSectors2` method.

Pre-commit deal IDs may be removed in a future FIP when deprecating the old methods.

##### SectorOnChainInfo

The per-sector `DealIDs` field is deprecated and to be ignored.
The miner actor will not write to it.

The `DealWeight` field is interpreted to carry the weight of _any non-zero, unverified data_ in a sector,
not only the weight of deals made via the built-in market actor.
Similarly, the `VerifiedDealWeight` field carries the weight of _any verified data_, regardless of
whether a "deal" was made via the built-in market actor.

Per-sector deal IDs may be removed from state in a future migration.

#### Sector activation

The miner actor exports new methods for sector activation:

> `ProveCommitSectors2` (method 33) and `ProveReplicaUpdates3` (method 34).

Reviewer: Now ProveReplicaUpdates2

    RequireActivationSuccess: bool,
    // Whether to abort if any notification returns a non-zero exit code.
    RequireNotificationSuccess: bool,
}

// Data to activate a commitment to one sector and its data.
// All pieces of data must be specified, whether or not not claiming a verified allocation or being
// notified to a data consumer.
// An implicit zero piece fills any remaining sector capacity.
struct SectorActivationManifest {
    // Sector to be activated.
    Sector: SectorNumber,
    // Pieces comprising the sector content, in order.
    Pieces: []PieceActivationManifest,
}

struct PieceActivationManifest {
    // Piece data commitment.
    CID: Cid,
    // Piece size.
    Size: PaddedPieceSize,
    // Identifies a verified allocation to be claimed.
    VerifiedAllocationKey: Option<VerifiedAllocationKey>,
    // Synchronous notifications to be sent to other actors after activation.
    Notify: []DataActivationNotification,
}

struct VerifiedAllocationKey {
    Client: ActorID,
    ID: AllocationID,
}

struct DataActivationNotification {
    // Actor to be notified.
    Address: Address,
    // Data to send in the notification.
    Payload: []byte,
}

> struct ProveCommit2Return {

Reviewer: We've changed this to 
```
pub struct ProveCommitSectors2Return {
    pub activation_results: BatchReturn,
}
```

is the same as for `ProveCommitSectors2`.

In execution of `ProveReplicaUpdates3`, the miner actor:

1. validates that the existing sectors are eligible for update;
2. computes an unsealed CID from the piece manifests;
3. verifies the sector update proofs or aggregate proof, and hence the new sealed and unsealed CIDs
   match the values declared and computed respectively;
4. claims any verified allocations specified in the piece manifest;
5. computes weights, pledge, power etc and activates the new sectors in state;
6. notifies any actors specified in the piece manifests.

```
struct ProveReplicaUpdates3Params {
    SectorUpdates: []SectorUpdateManifest,
    // Proofs for each sector, parallel to activation manifests.
    // Exactly one of sector_proofs or aggregate_proof must be non-empty.
    SectorProofs: [][]byte,
    // Aggregate proof for all sectors.
    // Exactly one of sector_proofs or aggregate_proof must be non-empty.
    AggregateProof: []byte,
    // The proof type for all sector update proofs, individually or before aggregation.
    UpdateProofsType: RegisteredUpdateProof,
    // The proof type for the aggregate proof (ignored if no aggregate proof).

>     AggregateProofType: RegisteredAggregateProof,

Author: @cryptonemo can you verify that this is the correct information that will need to be provided by an SP to submit an aggregated replica update proof?

Reviewer 2: This looks correct to me, thanks!  Still working on that FIP by the way ... :hourglass_flowing_sand:

    MinimumCommitmentEpoch: ChainEpoch,
    // Information about some pieces added to (or retained in) the sector.
    // This may be only a subset of sector content.
    // Inclusion here does not mean the piece was definitely absent previously.
    // Exclusion here does not mean a piece has been removed since a prior notification.
    Added: []PieceChange,
}

// Description of a piece of data committed to a sector.
struct PieceChange {
    Data: Cid,
    Size: PaddedPieceSize,
    // A receiver-specific identifier.
    // E.g. an encoded deal ID which the provider claims this piece satisfies.
    Payload: []byte],
}

// For each piece in each sector, the notifee returns an exit code and
// (possibly-empty) result data.
// The miner actor will pass through results to its caller.
// Note transparent serialization of single-element struct.
struct SectorContentChangedReturn {
    // A result for each sector that was notified, in the same order.
    Sectors: []]SectorReturn,
}

// Note transparent serialization of single-element struct.
struct SectorReturn {
    // A result for each piece for the sector that was notified, in the same order.

>     Added: []]PieceReturn,

Reviewer: typo: []PieceReturn

}

> struct PieceReturn {

Reviewer: We've changed this to

```
pub struct PieceReturn {
    // Indicates whether the receiver accepted the notification.
    // The caller is free to ignore this, but may chose to abort and roll back.
    pub accepted: bool,
}
```

aborting the operation if one fails.

**Sector activation includes claiming of any verified allocations**.
If a claim fails for one piece, no claims will be made for any piece in the sector and the sector will not be activated.
An SP cannot choose to have activation proceed despite an invalid claim;
they can instead resubmit the failed sector with only the remaining valid claims.

**Sector activation does not include notifications**.
Notifications are sent strictly _after_ activation, and only for successfully activated sectors.
If a notification call returns a non-zero exit code, sector activation will be committed regardless.
An SP can specify `RequireNotificationSuccess=true` to instead require every notification to succeed,
aborting the entire operation if one fails.
There is no ability to roll back the activation of only some sectors in response to failed notifications.

This is a change in sequencing and semantics from the existing onboarding methods flow,
where deal activation happens first and failure leads to an individual sector failing,
and verified claims happen second and must all succeed.

#### ProveCommitSector

> 

Reviewer: nit: good to mention ProveCommitAggregate here too

Reviewer: Probably also ProveReplicaUpdate

## Product Considerations
<!--All FIPs must contain a section that discusses the product implications/considerations relative to the proposed change. Include information that might be important for product discussion. A discussion on how the proposed change will enable better storage-related goods and services to be developed on Filecoin. FIP submissions missing the "Product Considerations" section will be rejected. An FIP cannot proceed to status "Final" without a Product Considerations discussion deemed sufficient by the reviewers.-->

## Implementation
<!--The implementations must be completed before any core FIP is given status "Final", but it need not be completed before the FIP is accepted. While there is merit to the approach of reaching consensus on the specification and rationale before writing code, the principle of "rough consensus and running code" is still useful when it comes to resolving many discussions of API details.-->

## TODO
- Determine whether we will migrate to remove SectorOnChainInfo.Deals, or merely deprecate its use.

> - Determine whether to deprecate ProveReplicaUpdates2 (and claim its name for PRU3)

Reviewer 3: Can we add a todo for "gas cost benchmarking and impact"?

Reviewer 4: Resolved in https://github.com/filecoin-project/FIPs/blob/987b3f8c2b36d0ea886ddaae8a4d986ad911c3a8/FIPS/fip-0076.md?plain=1#L691

---
fip: "<to be assigned>"
title: Direct data onboarding
author: "Alex North (@anorth), @zenground0"
discussions-to: https://github.com/filecoin-project/FIPs/discussions/730
status: Draft
type: Technical
category: Core
created: 2023-08-24
---

## Simple Summary

> Adds new `ProveCommitSectors2` (method 33) and `ProveReplicaUpdates2` (method 34) methods to the miner actor

Reviewer 3: ```suggestion
Add new `ProveCommitSectors2` (method 33) and `ProveReplicaUpdates2` (method 34) methods to the miner actor
```

Author: I will decline these suggestions. I have used a consistent declarative rather than imperative posture and I don't think it detracts from understanding the content.

> Adds new `ProveCommitSectors2` (method 33) and `ProveReplicaUpdates2` (method 34) methods to the miner actor

Author: Can you suggest some names? I've been unable to find something appropriately concise yet more descriptive of the difference.

> Adds new `ProveCommitSectors2` (method 33) and `ProveReplicaUpdates2` (method 34) methods to the miner actor

Reviewer 5: Yeah, I didn't have any good ideas either :(

> which support committing to data and claiming verified allocations without requiring a built-in market deal.

Reviewer 3: ```suggestion
which support committing data to sectors and claiming datacap allocations without requiring a built-in market (f05) deal.
```

> Adds a sector→deal mapping in the built-in market actor and deprecates the storage of deal IDs in the miner actor.

Reviewer 3: ```suggestion
Add a provider->sector→deal mapping in the built-in market actor and deprecated the storage of deal IDs in the sector onchain info of the miner actor.
```

Removes the legacy `PreCommitSector` (method 6) and `PreCommitSectorBatch` (method 25)
in favour of the existing `PreCommitSectorBatch2` (method 28) which requires
the sector unsealed CID to be specified while pre-committing.

> Removes the legacy and unused `ProveReplicaUpdates2` (method 29) (giving that name to the new method 34).

Reviewer 3: Peer: Ideally I would prefer a new method number to be assigned, and deprecate the old message/drop the code an upgrade after. Imagine a user sends legacy PRU2 message right before the upgrade, the message will fail upon block inclusion if it happens right after the network upgrade. It is a failure we could avoid for users with minimum cost imho.

Author: See design rationale. It's just not going to happen - it has never been used.

Existing onboarding flows that use the built-in market actor remain fully supported, but optional.

## Abstract

The only mechanism today by which storage providers (SPs) are permitted to commit data in sectors is

> to make a deal with the built-in storage market actor.

Reviewer 3: ```suggestion
to publish a deal made with data client to the built-in storage market actor.
```

The built-in market actor is expensive in gas, and offers only basic functionality.

> Requiring its use raises costs and limits the utility and value of sectors,

Reviewer 3: ```suggestion
Requiring its use raises unnecessary costs for those who don't use f05 market functionalities and limits the utility and value of sectors,
```

and the applications that can be built on Filecoin.

This cost and inconvenience is unnecessary in the common cases of verified deals with no client payments,
and storage arrangements that do not require any on-chain deal.

> The verified registry actor already records deal-like information describing DataCap allocation terms.

Reviewer 3: This sentence on it now makes sense, tho im not really sure what it tries to say within the context of this paragraph.  Could the follow be a bit clearer?
```suggestion
The other thing f05 deal offers is on-chain deal term, and for verified pieces, the verified registry actor already records deal-like information describing DataCap allocation terms.
```

This proposal adds new onboarding methods which support direct commitment of data into sectors,
without necessary reference to any deal.
This provides gas-cheap data onboarding for many use cases.

> It also introduces a new scheme for deal activation which can support deals (and other data-application transactions)

Reviewer 3: do we want to keep use the term `deal`? o should we start to move towards something like  `data (piece) activation`?

Author: The scheme right now is for the built-in market actor, so I think we should keep using that term here. The code doesn't though.

> It also introduces a new scheme for deal activation which can support deals (and other data-application transactions)

Reviewer 3: ```suggestion
It also introduces a new scheme in storage miner actor for deal activation which can support deals (and other data-application transactions)
```

> brokered by user-programmed smart contracts.

Reviewer 3: ```suggestion
brokered by user-programmed actors.
```

Author: I get what you're saying here, but the FVM doesn't currently permit user-programmed actors, and 😭 may never. They use the term "smart contract" to refer to a thing running inside another (built-in) actor, like EVM. That is realistically the way this will first be usable.

This new scheme is initially limited to the built-in market actor, but can be extended to other actors in the future.

## Change Motivation

The built-in market actor is expensive to use but provides very little utility to majority data onboarding use cases.

> A vast majority of deals today are simple FIL+ verified deals with no client payment.

Reviewer 3: ```suggestion
A vast majority of deals today are simple Fil+ verified deals with no client payment.
```

Author: They're forced to use collateral, so we can't say that yet.

Since the verified registry actor already records all relevant information about DataCap allocation terms
(the client and provider, piece commitment, and duration),
the built-in market actor's replication of this data is unnecessary.
The built-in market plays no necessary role in allocating or accounting for QA power.

> Publishing deals is the largest on-chain cost of QA power onboarding,

Reviewer 3: isn't PSD also costy for regular deals too?

currently consuming a large fraction (about half) of total chain bandwidth.
If the built-in storage market actor were bypassed for the common cases of un-paid or off-chain settled deals,
both verified and unverified data could be onboarded at a significant reduction in gas cost.

The built-in market actor is also very limited.
It supports only a single, simple deal schema and policies, and does not support a range of desirable market features.
User-programmed smart contracts are restricted in the functionality that they can provide by the necessity
of using the built-in market actor as an intermediary for any data commitments.
They are also restricted by the lack of any hooks into the built-in miner actor to be notified of sector data
commitments.

> Requiring the built-in market actor raises costs and limits the utility and value of sectors,

Reviewer 3: ➕

and the applications that can be built on Filecoin.
It cannot support the expansion in usage or functionality that network participants hope for.

This proposal aims to:

> - Support data onboarding, including Fil+ verified data, with no smart-contract intermediary;

Reviewer 3: ```suggestion
- Support data onboarding, including Fil+ verified data, with no built-in storage market or user deployed actor intermediary;
```

> - Provide a new scheme for data activation which can later support

Reviewer 3: ```suggestion
- Provide a new scheme for data activation triggered by storage miner actors which can later support
```

  user-programmed smart contracts to function as data storage applications (including as markets);

> - Continue supporting existing onboarding methods to give participants time to migrate their workflows.

Reviewer 3: ```suggestion
- Continue supporting existing onboarding methods to keep existing f05 storage markets flow and functionality remain supported (for clients who would like to have onchain payments) and give participants time to migrate their workflows.
```

Author: Parties can have on-chain payments without using the old methods. The old methods remain only for transition.

Reviewer 3: > Parties can have on-chain payments without using the old methods

I understand - however, I think it is too soon to assume all network users want to migrate to the new flow. if we can have a quick mention on if one would like keep their exact existing flow - how should they do so might be useful. i.e: PSD -> PreCommitBatch2 -> ProveCommitSectors/ProveCommitAggregation; or PSD -> PreCommitBatch2 -> ProveCommitSectors2 with RequireNotificationSuccess=true && RequireActivationSuccess==true

## Specification

### Overview

Direct data onboarding comprises changes to the built-in miner and market actors
to provide new onboarding methods that do not require a built-in market deal.
These new methods accept additional parameters from the storage provider which specify
the pieces of data activated in the sector, any verified allocations being claimed,

> and the address of any actors to notify about the successful activation.

Reviewer 3: ```suggestion
and the address of any actors to notify about the successful data storage activation.
```

> Deals are possible, but not necessary.

Reviewer 3: ```suggestion
Builtin storage market (f05) deals are possible, but not necessary.
```

In the direct onboarding flow:

> - _[for verified data only]_ a client makes a verified allocation directly with the verified registry

Reviewer 3: ```suggestion
- _[for verified data only]_ a verified client makes a datacap allocation for a specific piece of date to be stored with a designated storage provider directly with the verified registry
```

Author: If you don't think it's necessary to specify that a deal is for specific data and a designated provider also in the line below, I don't think it's necessary to specify here either. How allocations work is already specified in other FIPs and not changing here.

>   by transferring DataCap tokens to it (this is already possible today),

Reviewer 3: Peer: should we mention verified client can also delegate this operation a trusted third party, like SPs, or DP, where trusted third party may allocate datacap on behalf of the client by traversing datacap to verifreg via `transfer_from`?

- _[for on-chain paid deals only]_ an SP publishes storage deals to the built-in market actor (also already possible
  today),
- at pre-commit, an SP must specify a sector’s data commitment (unsealed CID),
  but does not need to specify the structure of that data nor any deals or verified allocations.
- at prove-commit or replica-update, an SP specifies the pieces of data comprising a sector,
  and for each piece may nominate:
    - a verified allocation to claim, and/or
    - an actor to notify of the commitment (e.g. to activate a deal).

The miner actor verifies that the pieces of data claimed by the SP correspond to the data commitment proven.
The verified registry is involved only if verified data is being claimed.

> The built-in market actor is involved only if an on-chain-settled deal is required (i.e. if non-zero payment).

Author: Changed to "need only be involved". An SP is not really using direct onboarding if they do a deal – I think direct _means_ no deal. But the new methods also support doing deals, and I can see how a reader might think direct == new methods rather than direct == no deal. Such definition isn't relevant to the protocol though.

#### State

A new field maps sector numbers to the deal IDs that the market has been notified are stored in those sectors.

```
// New structure storing per-sector deal information.
struct SectorDeals {
    Deals: []DealID
}

// Existing deal state structure gets a new field with sector number.
struct DealState {
    // All existing fields as today.
    // ...

    // 0 if not yet included in proven sector (0 is also a valid sector number).
    SectorNumber: SectorNumber,
}

struct State {
    // All existing state as today.
    // ...

    // Existing mapping of deal state by ID.
    // Since deal state now includes sector number, this gives deal->sector index.
    // AMT[DealID]DealState
	States: Cid 

    // New mapping of sector IDs to deal IDs, grouped by storage provider.
    // HAMT[Address]HAMT[SectorNumber]SectorDeals
    ProviderSectors: Cid 
}
```

#### BatchActivateDeals

When an SP activates a piece with the existing activation methods,
deals are activated with the `BatchActivateDeals` (method 6).
This method returns the necessary verified allocation IDs to the miner actor.
The `BatchActivateDeals` method parameters are expanded to include the sector ID for each deal.

> When activating a deal, the market actors writes it into the new `ProviderSectors` mapping.

Reviewer 5: ```suggestion
When activating a deal, the market actor writes it into the new `ProviderSectors` mapping.
```

This proposal deprecated the miner methods `PreCommitSector` (method 6), `PreCommitSectorBatch` (method 25),
and `ProveReplicaUpdates2` (method 29).
All three have alternatives already available on mainnet that should be used instead.

This proposal requires a state migration to the market actor to add the new `ProviderSectors` mapping.
Computing this mapping requires reading all sector metadata from the miner actor.

This proposal requires a network upgrade to deploy the new built-in actor code.

## Test Cases

To be provided with implementation (see below).

## Security Considerations

This proposal has no impact on consensus or blockchain security.

## Incentive Considerations

<!--All FIPs must contain a section that discusses the incentive implications/considerations relative to the proposed change. Include information that might be important for incentive discussion. A discussion on how the proposed change will incentivize reliable and useful storage is required. FIP submissions missing the "Incentive Considerations" section will be rejected. An FIP cannot proceed to status "Final" without a Incentive Considerations discussion deemed sufficient by the reviewers.-->

> - gas impacts

Reviewer 5: TODO

Author: already done below, this was stray. But now filling out other incentive considerations.

As compared with current workflow costs:

- For direct onboarding with no verified claim, neither publishing a deal nor transferring datacap are necessary.
  Thus, the full cost of `PublishStorageDeals` is avoided.
  The cost of sector pre-commitment is reduced slightly by avoiding interaction with the built-in market actor
  The cost of sector activation is reduced by avoiding the cost of deal activation.
- For direct onboarding with a verified claim, most of the cost of `PublishStorageDeals` is avoided.
  The cost of transferring DataCap is still incurred, but is only a fraction of the cost of publishing a deal.
  The cost of sector pre-commitment is reduced slightly by avoiding interaction with the built-in market actor
  The cost of sector activation is reduced by avoiding the cost of deal activation.
- For direct onboarding with a verified claim and a built-in storage market deal, the cost of `PreCommitSectorBatch`
  is reduced slightly by avoiding interaction with the built-in market actor.

As an illustrative example, an analysis of the gas cost of selected onboarding messages
with a batch size of 16 in epoch 2844239 indicates:

> | Message              | Observed w/ Fil+ | Direct with Fil+ | Observed ex Fil+ | Direct ex Fil+   | 

Reviewer 3: is ex FIL+ for both CC sectors and sectors only has regular deals?

Author: It's not for CC since they never interact with the market or verifreg anyway. This proposal will make no significant difference to CC gas cost.

the total amount and composition of data and verified data in the network.

After this proposal, information representing data commitments, verified claims, and deals are maintained separately.

- The source of truth for sector data commitments is the piece activation manifests in the message history.
- The source of truth for verified data claims is the verified registry actor state.
- The built-in market actor state is a source of truth only for on-chain deal accounting.

The built-in market actor may have a partial view of both piece commitments and verified claims,
but this is no longer a reliable or complete source of information about either.
Clients, explorers and analytic tools should inspect the source of truth for each type of information,
and avoid reliance on the presence of deals with built-in market actor.

## Implementation

Implementation of the protocol changes is being developed in the `integration/direct-onboarding`
[branch of the built-in actors](https://github.com/filecoin-project/builtin-actors/tree/integration/direct-onboarding)
repository.

## TODO

> - Determine whether we will migrate to remove SectorOnChainInfo.Deals, or merely deprecate its use.

Reviewer 5: Primary motivation to do so: we're migrating all sectors this upgrade anyway, now's a good time to cleanup.

Primary motivation not to do so: it's a change (albeit a small one) that will save very little state size. Is it worth it?

Author: Have decided not to remove sector deals immediately.
